jobs/update_orphaning_dashboard_etl.py

```diff
@@ -0,0 +1,1236 @@
+    query_job = bq.query(longitudinal_shim_sql, job_config=job_config)
+    # Wait for query execution
+    query_job.result()
+    print("BigQuery aggregation finished")
+
+
+def longitudinal_shim_transform(project, dataset, table):
+    """Transform preaggregated dataset to match legacy Longitudinal schema and make it available as
+     `longitudinal_shim` in SQL context.
+    """
+    # BigQuery Storage API can't handle the schema and queries (`out_of_date_details_sql`) we have here
+    # As a workaround, we'll load AVRO-exported dump to Spark
+    print("Exporting BigQuery table as AVRO...")
+    destination_uri = GCS_TABLE_DUMP_PATH + "*"
+
+    job_config = bigquery.ExtractJobConfig()
+    job_config.destination_format = "AVRO"
+
+    bq = bigquery.Client()
+    table_ref = bq.dataset(dataset, project=project).table(table)
+
+    extract_job = bq.extract_table(
+        table_ref,
+        destination_uri,
+        location="US",
+        job_config=job_config
+    )  
+    extract_job.result()  # Waits for job to complete.
+
+    print(
+        "Exported {} to {}".format(table_ref, destination_uri)
+    )
+
+    longitudinal_shim_raw_from_avro=spark.read.format("avro").load(GCS_TABLE_DUMP_PATH)
+
+    def flatten_settings_update(settings):
+        channel_list = list(map(lambda x: x[0], settings["update"]["channel"]))
+        enabled_list = list(map(lambda x: x[0], settings["update"]["enabled"]))
+        return {"update": {"channel": channel_list, "enabled": enabled_list}}
+
+    return_struct_type = StructType([StructField("update", StructType([StructField("channel", ArrayType(StringType())), StructField("enabled", ArrayType(BooleanType()))]))])
+    udf_flatten_settings_update = udf(flatten_settings_update, return_struct_type)
+
+    def fix_schema(raw_longitudinal_df):
+        with_settings_fixed = raw_longitudinal_df.withColumn("settings_fixed", udf_flatten_settings_update("settings")).drop("settings").withColumnRenamed("settings_fixed", "settings")
+        return with_settings_fixed
+
+
+
+    longitudinal_shim_raw_from_avro_fixed = fix_schema(longitudinal_shim_raw_from_avro)
+
+
+
+
+    def merge_keyed_count_histograms(histograms, histogram_name):
+        keys = []
+        for histogram_struct in histograms:
+            histogram_array = histogram_struct[histogram_name]
+            if histogram_array:
+                for key, count_histogram_string in histogram_array:
+                    if key not in keys:
+                        keys.append(key)
+
+        res = {}
+        n_hist = len(histograms)
+        for key in keys:
+            res[key] = [0]*n_hist
+            for i, histogram_struct in enumerate(histograms):
+                histogram_array = histogram_struct[histogram_name]
+                if histogram_array:
+                    for hist_key, count_histogram_string in histogram_array:
+                        if hist_key == key:
```

Comment:
Why not move this to the earlier iterator? Here we're iterating over all histograms to get all the keys; then for each key, iterating over all the histogram's keys again. Seems we could do it in one pass, and this would be clearer.